### PR TITLE
Command Migration: ('HINCRBY', 'HINCRBYFLOAT', 'HRANDFIELD')

### LINKED ISSUE #1023: Command Migration: ('HINCRBY', 'HINCRBYFLOAT', 'HRANDFIELD')
This issue tracks the migration of the mentioned commands - ('HINCRBY', 'HINCRBYFLOAT', 'HRANDFIELD') to make them compatible across the three protocols supported by the Dice database: RESP, HTTP, and WebSocket.
DiceDB now supports more than one protocols (Resp/http/websocket). We don't want eval operation of each command to be strictly bound with any of the protocols. Currently eval function return statements of each command is bounded to the RESP protocol. 

*The migration is required because:*

- The current implementation is specific to the RESP protocol
- Evaluation functions of each command must be refactored to ensure protocol-agnostic behavior.
- Evaluation functions should return raw responses instead of protocol specific responses.

The goal is to make the command logic protocol-independent, allowing all three protocols to call the same core functionality seamlessly.

### Requirements

- Refactor the evaluation function of `[Command Name]` to be generic.
- Ensure the evaluation function does not include protocol-specific logic.

### Migration Steps

1. **Analyze Current Implementation**
   - Review the current codebase to understand how the command logic is implemented.

2. **Refactor Return Logic**
   - Create a new function with the function definition as `evalXXX(args []string, store *dstore.Store) *EvalResponse` under the file `/internal/eval/store_eval.go`
   - Analyse return statements of the eval and modify them to send raw types without encoding
   - Use errors in the return statements from `/errors/migrated_errors.go` file
   - Use perdefined responses from `/internal/clientio/resp.go` file as RespType variables

3. **Command/Worker specific Changes**
   - Make `IsMigrated` flag to true in the commands information under `/internal/eval/commands.go` file.
   - Use the newly written eval function against the `NewEval` parameter in the same command structure.
   - Delete the old eval function from the `/internal/eval/eval.go` file.
   - Add the migrated command to the `/internal/worker/CommandsMeta` map and make it's type as `SingleShard`

4. **Update Unit Tests**
   - Refactor existing unit tests to accommodate the new implementation.
   - Add new unit tests if necessary to cover all possible cases.

5. **Integration Tests**
   - Run all integration tests to ensure successful migration.
   - Ensure that each protocol (RESP, HTTP, WebSocket) works correctly after migration.

### Checklist

- [ ] Migrated the evalXXX function with the latest definition
- [ ] Update or add unit tests for the new implementation.
- [ ] All unit tests pass successfully.
- [ ] Ensure all integration tests pass successfully.

### Additional Notes

- Describe any edge cases that need to be handled in the generic function.
- Mention any protocol-specific optimizations required in wrappers.
  
If there are any questions or concerns about this migration, please mention them here.

### Related Issues/PRs

- Sample implementation for the Get, Set, GetSet and SetEx can be found in [this Pull Request.](https://github.com/DiceDB/dice/pull/766)

### PR DESCRIPTION
**Pull request by sashpawar11**
This PR migrates HINCRBY, HINCRBYFLOAT and HRANDFIELD to make it protocol agnostic. Closes #1023 

Migration Checklist

- [x]  Migrated the evalXXX function with the latest definition
- [x]  Update or add unit tests for the new implementation.
- [x]  All unit tests pass successfully.
- [x]  Ensure all integration tests pass successfully.
- [x] Ensure integration tests are added for the migrated command on multi-threaded resp server.
- [x] Add http, websocket tests for the commands
- [x] Move Integration tests for the respective commands under the RESP integration tests directory from Async directory
- [x] Please validate that the documentation for the respective commands is up to date. If not then consider adding them. 

Note:
 
- Http, Resp and Websocket tests to be added as well since they are missing for the addressed commands under this issue.
- Docs for these commands have been added

### CODE REVIEW

**Comment by sashpawar11** [comment]
Hi @AshwinKul28 
update 10/15 - unit tests have been migrated, please have a look and provide feedback! The unit tests have passed succesfully.

Will continue to work on migrating the integration tests.
Thankyou.

**Comment by sashpawar11** [comment]
@AshwinKul28 - the HRANDFIELD depends on selectRandomFields() for returning the random field. 
Should we move it to hash.go which feels more relevant as it has other utility functions around hashmaps? 
Currently, the selectRandomFields() has been migrated to store_eval.go from eval.go under this PR

**Comment by sashpawar11** [comment]
Hi @AshwinKul28 - for the last task, I'm not sure how to proceed on it.
Are these commands optimized for multi-thread execution yet?

The unit tests and integration tests have passed in the latest build

**Comment by soumya-codes** [comment]
@sashpawar11 please refer to existing integration tests under integration_tests/commands/resp for the commands that are already migrated.
We will need to do the same these commands as well.

**Comment by sashpawar11** [comment]
@soumya-codes Awesome, thanks!! Will have a look at it and migrate these commands.
Thankyou

**Comment by sashpawar11** [comment]
Update 10/16 - added migrated resp tests for the commands. All the tasks have been completed and build passed, however, the http and websocket tests are not present for these commands. Added an additional task in the description for the same, will complete it by today/tomorrow and update here once ready for review.

cc @AshwinKul28 @soumya-codes

**Comment by sashpawar11** [comment]
Update 10/18 - Added all the missing HTTP integrations tests for the command. Will try complete the web-socket tests asap and mark this PR for review.
cc @AshwinKul28 @soumya-codes

File: integration_tests/commands/async/hincrby_test.go
Comment (lucifercr07) [inline integration_tests/commands/async/hincrby_test.go:1]: We can remove these tests now as we've migrated.

**Review by lucifercr07 — APPROVED** [review]

**Review by lucifercr07 — CHANGES_REQUESTED** [review]

File: integration_tests/commands/async/hincrbyfloat_test.go
[the remark this answers is not in the document]
Comment (lucifercr07) [reply integration_tests/commands/async/hincrbyfloat_test.go:1]: HTTP websockets is required, only the ones inside the `async` folder.

[the remark this answers is not in the document]
Comment (sashpawar11) [reply integration_tests/commands/async/hincrbyfloat_test.go:1]: Oh cool got it, my bad. Thanks @lucifercr07 - will removet the async tests.

[the remark this answers is not in the document]
Comment (sashpawar11) [reply integration_tests/commands/async/hincrbyfloat_test.go:1]: Commit added to remove the async tests.

**Comment by sashpawar11** [comment]
Hi @AshwinKul28 @soumya-codes @lucifercr07 - all the commands have been migrated and their respective tests have been added. Resolved multiple merge conflicts along the way. 

The PR Is ready for review.


The documentation for the commands are missing and I'll try to add it via a separate PR as this one is already a bit large.
Thankyou.

**Comment by sashpawar11** [comment]
> @AshwinKul28 - the HRANDFIELD command depends on `selectRandomFields()` for returning the random field. Should we move it to `hash.go` which feels more relevant as it has other utility methods around hashmaps? Currently, `selectRandomFields()`has been migrated to `store_eval.go` from `eval.go` under this PR

Also please advise on this.
cc @AshwinKul28 @soumya-codes @lucifercr07

File: integration_tests/commands/async/hincrbyfloat_test.go
[the remark this answers is not in the document]
Comment (lucifercr07) [reply integration_tests/commands/async/hincrbyfloat_test.go:1]: @sashpawar11 is this missed in commit?

**Review by lucifercr07 — CHANGES_REQUESTED** [review]
Can we also check the doc related to `'HINCRBY', 'HINCRBYFLOAT', 'HRANDFIELD'` commands also once, if there's any discrepancy?

Minor comments added.

**Comment by sashpawar11** [comment]
> Can we also check the doc related to `'HINCRBY', 'HINCRBYFLOAT', 'HRANDFIELD'` commands also once, if there's any discrepancy?
> 
> Minor comments added.

Hey @lucifercr07 - there no docs present for these commands at the moment. Will be adding them in a seperate pr as I complete working on it.

File: internal/eval/store_eval.go [hunk 1]
```diff
@@ -910,3 +912,214 @@ func evalPFMERGE(args []string, store *dstore.Store) *EvalResponse {
...
+			}
+			withValues = true
+		}
+	}
+
+	return selectRandomFields(hashMap, count, withValues)
+}
+
+// selectRandomFields returns random fields from a hashmap.
+func selectRandomFields(hashMap HashMap, count int, withValues bool) *EvalResponse {
```
[the remark this answers is not in the document]
Comment (sashpawar11) [reply internal/eval/store_eval.go:1083]: Yep, will migrate it to hmap.go, had added a question regarding it on this PR initially.
Thanks.

**Comment by sashpawar11** [comment]
Hi @AshwinKul28 @soumya-codes @lucifercr07 - all the tasks have been completed. Please review the changes!

Latest changes:
- Documentation for the commands have been added as well.
- `selectrandomfields() `has been moved over to `hmap.go`

**Comment by AshwinKul28** [comment]
Hey @sashpawar11 this is a commendable effort. Everything looks great to me. Few cosmetic changes are required. (I know this is a little back and forth, apologies for that. But we want to make it a one-time effort, and some ideal PRs to merge)

**Comment by sashpawar11** [comment]
Thanks @AshwinKul28 ! And yes I agree, lets complete all the changes to make this a one-time effort

**Review by AshwinKul28 — CHANGES_REQUESTED** [review]

File: internal/eval/store_eval.go [hunk 2]
```diff
@@ -453,69 +453,6 @@ func evalZRANGE(args []string, store *dstore.Store) *EvalResponse {
 	}
 }
 
-// evalZRANK returns the rank of the member in the sorted set stored at key.
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:456]: Hi @sashpawar11 I see this is mistakenly got removed.

File: internal/eval/store_eval.go [hunk 3]
```diff
@@ -996,6 +933,190 @@ func evalPFMERGE(args []string, store *dstore.Store) *EvalResponse {
...
+		return &EvalResponse{
+			Result: nil,
+			Error:  diceerrors.ErrWrongTypeOperation,
+		}
+	}
+
+	hashMap := obj.Value.(HashMap)
+	if len(hashMap) == 0 {
+		return &EvalResponse{
+			Result: clientio.RespEmptyArray,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:1086]:
```suggestion
			Result: clientio.EmptyArray,
```

File: docs/src/content/docs/commands/ZRANK.md
Comment (AshwinKul28) [inline docs/src/content/docs/commands/ZRANK.md:1]: HI @sashpawar11 I see this file also got removed mistakenly. Please revert.

File: internal/eval/store_eval.go (hunk 2 above)
Comment (sashpawar11) [reply internal/eval/store_eval.go:456]: Gosh, my bad - might have missed it during conflict resolution. Will fix and add it back soon

File: docs/src/content/docs/commands/ZRANK.md
Comment (sashpawar11) [reply docs/src/content/docs/commands/ZRANK.md:1]: Yep, the tests have also disappeared , will add them back from the commit - apologies

**Comment by AshwinKul28** [comment]
Thanks @sashpawar11 Please pull the latest from the master, I see ZRANK command implementation and documentation got removed in your PR. Please rebase with the master. Everything else looks great. Thanks again!

**Comment by sashpawar11** [comment]
> Thanks @sashpawar11 Please pull the latest from the master, I see ZRANK command implementation and documentation got removed in your PR. Please rebase with the master. Everything else looks great. Thanks again!

Fixed , ZRANK is back. Ready for merge.
Thankyou.

**Comment by AshwinKul28** [comment]
Ahh, another PR got merged. 🗡️  Please resolve conflicts quickly @sashpawar11 and I will merge. Thanks

**Review by AshwinKul28 — APPROVED** [review]
LGTM

**Comment by sashpawar11** [comment]
> Ahh, another PR got merged. 🗡️  Please resolve conflicts quickly @sashpawar11 and I will merge. Thanks

Thanks for resolving!

**Comment by sashpawar11** [comment]
> Ahh, another PR got merged. 🗡️  Please resolve conflicts quickly @sashpawar11 and I will merge. Thanks

Hey, will try to resolve soon

**Comment by sashpawar11** [comment]
Conflicts resolved. Ready for merge.
Thankyou
